db/src/db.rs

```diff
@@ -17,9 +20,135 @@ pub struct RocksDB {
     pub(crate) inner: Arc<OptimisticTransactionDB>,
 }
 
+trait RocksDBOptionsConversion: Sized {
+    fn convert(value_str: &str) -> StdResult<Self, ()>;
+}
+
+impl<T> RocksDBOptionsConversion for T
+where
+    T: FromStr,
+{
+    fn convert(s: &str) -> StdResult<Self, ()> {
+        Self::from_str(s).map_err(|_| ())
+    }
+}
+
+macro_rules! set_option {
+    ($opts:ident, $map:ident, $func:ident, $key:literal, $type:ty) => {
+        if let Some(value) = $map.remove($key) {
+            let v: $type = RocksDBOptionsConversion::convert(&value).map_err(|_| {
+                internal_error(format!(
+                    "failed to parse value of database option \"{}\"",
+                    $key
+                ))
+            })?;
+            $opts.$func(v);
+        }
+    };
+}
+
+// Load options which are not dynamically changeable through SetDBOptions() API.
+fn load_non_dynamic_options(t: &mut HashMap<String, String>) -> Result<Options> {
```

Comment:
In rocksdb option file, these options are not in the same category as cfoptions:
https://github.com/facebook/rocksdb/blob/master/examples/rocksdb_option_file_example.ini#L36-L79

should we consider using another field?
```
pub struct Config {
    pub path: PathBuf,
    pub options: Option<HashMap<String, String>>,
    pub column_family_options: HashMap<String, HashMap<String, String>>,
}
```

Reply:
Column family options is supported in #2159.